File: lib/rgfa/lines/finders.rb
#
# Methods for the RGFA class, which allow to add lines.
#
# @tested_in api_lines_finders
#
module RGFA::Lines::Finders

  # @!macro [new] segment
  #   Searches the segment with name equal to +segment_name+.
  #   @param s
  #     [Symbol, String, RGFA::Line::Segment::GFA1, RGFA::Line::Segment::GFA2]
  #     segment name or instance
  #   @return [RGFA::Line::Segment::GFA1, RGFA::Line::Segment::GFA2]
  #     if a segment is found
  # @return [nil] if no such segment exists in the RGFA instance
  #
  def segment(s)
    return s if s.kind_of?(RGFA::Line)
    @records[:S][s.to_sym]
  end

  # @!macro segment
  # @raise [RGFA::NotFoundError] if no such segment exists
  def segment!(s)
    seg = segment(s)
    if seg.nil?
      raise RGFA::NotFoundError, "No segment has name #{s}"+
             "#{segment_names.size < 10 ?
               "\nSegment names: "+segment_names.inspect : ''}"
    end
    seg
  end

  # Record types whose references are stored in the RGFA instance
  # in hashes, where the key is a name field
  RECORDS_WITH_NAME = [:E, :S, :P, :U, :G, :O, nil]

  # Find the line with the given l
  # @param l [Symbol, RGFA::Line, RGFA::Placeholder] the line to search
  # @return [RGFA::Line, nil] if +l+ is a line, then it is returned;
  #   otherwise, the line is search with name +l+; if such a line does
  #   not exist +nil+ is returned
  def line(l)
    return nil if l.placeholder?
    case l
    when RGFA::Line
      return l
    when Symbol
      return line_by_name(l)
    when String
      return line_by_name(l.to_sym)
    else
      return nil
    end
  end

  # Find the line with the given name, and raise an exception if it does not
  # exist
  # @param l [Symbol, RGFA::Line, RGFA::Placeholder] the line to search
  # @raise [RGFA::ValueError] if +l+ is a placeholder
  # @raise [RGFA::NotFoundError] if no line with the given name +l+ exists
  # @return [RGFA::Line, nil] if +l+ is a line, then it is returned;
  #   otherwise, return the line with name +l+
  def line!(l)
    gfa_line = line(l)
    if gfa_line.nil?
      if l.placeholder?
        raise RGFA::ValueError,
          "Cannot search a line with l '*'"
      else
        raise RGFA::NotFoundError,
          "No line found with ID '#{l}'"
      end
    end
    return gfa_line
  end

  # Returns all the fragments where the ```external``` ID is the specified ID.
  # @return [Array<RGFA::Line::Fragment>]
  def fragments_for_external(id)
    @records[:F].fetch(id.to_sym, [])
  end

  # @overload select(hash)
  #   Searches a line based on the value of some of its fields
  #   @param hash [Hash] an hash field name => field content; if the key :name
  #     exists, the name field of the line type is searched first, then the
  #     other fields are compared; placeholder values are ignored.
  #   @return [Array<RGFA::Line>] a list of lines whose fields are compatible
  #     with the values given in +hash+
  # @overload select(line)
  #   Searches a line based on a given line instance
  #   @param line [RGFA::Line] a line
  #   @return [Array<RGFA::Line>] a list of lines with the same values in
  #     the mandatory fields and specified tags, as +line+; additional tags
  #     may be present; fields whose value is a placeholder are not compared
  def select(hash_or_line)
    is_hash = hash_or_line.kind_of?(Hash)
    name = is_hash ? hash_or_line[:name] : hash_or_line.get(:name)
    if !name.nil? and !name.placeholder?
      collection = [line_by_name(name)]
    else
      record_type = is_hash ? hash_or_line[:record_type] :
                              hash_or_line.record_type
      collection = collection_for_select(record_type)
    end
    method = is_hash ? :field_values? : :eql_fields?
    collection.select do |line|
      line.send(method, hash_or_line, [:record_type, :name])
    end
  end

  # @api private
  # @tested_in unit_lines_finders
  module API_PRIVATE

    # Search a possible duplicate of the gfa_line.
    # @param gfa_line [RGFA::Line]
    # @return [RGFA::Line, nil]
    def search_duplicate(gfa_line)
      case gfa_line.record_type
      when :L
        return search_link(gfa_line.oriented_from,
                           gfa_line.oriented_to, gfa_line.alignment)
      when *RECORDS_WITH_NAME
        return line(gfa_line.name)
      else
        return nil
      end
    end

    # Search the link from a segment S1 in a given orientation
    # to another segment S2 in a given, or the equivalent
    # link from S2 to S1 with inverted orientations.
    #
    # @param [RGFA::OrientedLine] oriented_segment1 a segment with orientation
    # @param [RGFA::OrientedLine] oriented_segment2 a segment with orientation
    # @param [RGFA::Alignment::CIGAR] cigar
    # @return [RGFA::Line::Edge::Link] the first link found
    # @return [nil] if no link is found.
    def search_link(oriented_segment1, oriented_segment2, cigar)
      s = segment(oriented_segment1.line)
      return nil if s.nil?
      s.dovetails.each do |l|
        return l if l.kind_of?(RGFA::Line::Edge::Link) and
          l.compatible?(oriented_segment1, oriented_segment2, cigar, true)
      end
      return nil
    end

  end
  include API_PRIVATE

  private

  # Searches a line by its name
  # @param name [Symbol]
  # @return [RGFA::Line, nil]
  def line_by_name(name)
    RECORDS_WITH_NAME.each do |rt|
      next if !@records[rt]
      found = @records[rt][name]
      return found if !found.nil?
    end
    return nil
  end

  # Computes a collection of lines compatible with the given
  # record_type.
  #
  # @param record_type [Symbol] The record type.
  def collection_for_select(record_type)
    case record_type
    when nil
      return lines
    when :S, :P
      collection = @records[record_type].values
    when :E, :G, :U, :O, :F
      collection = @records[record_type].values.flatten
    else
      collection = @records[record_type]
    end
  end

end

File: gfapy/lines/finders.py
import gfapy

class Finders:
  def segment(self, s):
    """Search a segment in a GFA.

    If the argument is a line, it is returned. If it is a string,
    it is used as a segment identifier, and the segment with that identifier
    is returned. If no segment has the identifier, None is returned.
    Note that None is also returned if a line of another type (not segment)
    has the identifier (differently from the line() method, which returns
    the lines independently from the record type).

    Parameters:
      l (str, gfapy.Line)
    """
    if isinstance(s, gfapy.Line):
      return s
    else:
      return self._records["S"].get(s, None)

  def try_get_segment(self, s):
    """Call segment() and raise an exception is the segment is not found."""
    seg = self.segment(s)
    if seg is None:
      raise gfapy.NotFoundError("No segment has name {}".format(s))
    else:
      return seg

  RECORDS_WITH_NAME = ["E", "S", "P", "U", "G", "O", "\n"]

  def line(self, l):
    """Search a line in a GFA.

    If the argument is a line, it is returned. If it is a string,
    it is used as a line identifier, and the line with that identifier
    is returned. If no line has the identifier, None is returned.

    Parameters:
      l (str, gfapy.Line)
    """
    if gfapy.is_placeholder(l):
      return None
    elif isinstance(l, gfapy.Line):
      return l
    elif isinstance(l, str):
      return self.__line_by_name(l)
    else:
      return None

  def try_get_line(self, l):
    """Call line() and raise an exception is the line is not found."""
    gfa_line = self.line(l)
    if gfa_line is None:
      if gfapy.is_placeholder(l):
        raise gfapy.ValueError(
          "'*' is a placeholder and not a valid name for a line")
      else:
        raise gfapy.NotFoundError(
            "No line found with ID {}".format(l))
    return gfa_line

  def fragments_for_external(self, external_id):
    """List of F lines with a given external ID."""
    return list(self._records["F"].get(external_id,{}).values())

  def select(self, dict_or_line):
    """Select all lines which respect a chriterion.

    The chriterion is expressed by the argument, which is either a line
    instance or a dictionary. If it is a dictionary, it shall contain
    pairs of fieldnames/values and the method returns all lines
    where the mentioned fieldnames have the corresponding values.
    If it is a line, it is compared with the lines of the same type
    in the Gfa instance and lines with the same field values
    are returned (undefined placeholder values are thereby not compared).
    """
    is_dict = isinstance(dict_or_line, dict)
    name = dict_or_line.get("name",None) if is_dict else dict_or_line.get("name")
    if name is not None and not gfapy.is_placeholder(name):
      collection = [self.__line_by_name(name)]
    else:
      if is_dict:
        record_type = dict_or_line.get("record_type",None)
      else:
        record_type = dict_or_line.record_type
      collection = self.__collection_for_select(record_type)
    method = "_has_field_values" if is_dict else "_has_eql_fields"
    return [line for line in collection \
        if getattr(line, method)(dict_or_line, ["record_type","name"])]

  def _search_duplicate(self, gfa_line):
    if gfa_line.record_type == "L":
      return self._search_link(gfa_line.oriented_from, gfa_line.oriented_to,
                               gfa_line.alignment)
    elif gfa_line.record_type in self.RECORDS_WITH_NAME:
      return self.line(gfa_line.name)
    else:
      return None

  def _search_link(self, orseg1, orseg2, cigar):
    s = self.segment(orseg1.line)
    if s is None:
      return None
    for l in s.dovetails:
      if isinstance(l, gfapy.line.edge.Link) and \
          l.is_compatible(orseg1, orseg2, cigar, True):
        return l
    return None

  def __line_by_name(self, name):
    for rt in self.RECORDS_WITH_NAME:
      if rt not in self._records:
        next
      found = self._records[rt].get(name, None)
      if found is not None:
        return found
    return None

  def __collection_for_select(self, record_type):
    if record_type is None:
      return self.lines
    else:
      d = self._records[record_type]
      if record_type == "F":
        retval = []
        for v in d.values():
          retval.extend(list(v.values()))
        return retval
      else:
        return list(d.values())
